warehouse/organizations/models.py

```diff
@@ -311,3 +341,140 @@ class OrganizationInvitation(db.Model):
+        ForeignKey("teams.id", onupdate="CASCADE", ondelete="CASCADE"),
+        nullable=False,
+    )
+
+    user = orm.relationship(User, lazy=False)
+    team = orm.relationship("Team", lazy=False)
+
+
+class TeamProjectRoleType(str, enum.Enum):
+
+    Admin = "Admin"
+    Upload = "Upload"
+
+
+class TeamProjectRole(db.Model):
+
+    __tablename__ = "team_project_roles"
+    __table_args__ = (
+        Index("team_project_roles_project_id_idx", "project_id"),
+        Index("team_project_roles_team_id_idx", "team_id"),
+        UniqueConstraint(
+            "project_id",
+            "team_id",
+            name="_team_project_roles_project_team_uc",
+        ),
+    )
+
+    __repr__ = make_repr("role_name", "team", "project")
+
+    role_name = Column(
+        Enum(TeamProjectRoleType, values_callable=lambda x: [e.value for e in x]),
+        nullable=False,
+    )
+    project_id = Column(
+        ForeignKey("projects.id", onupdate="CASCADE", ondelete="CASCADE"),
+        nullable=False,
+    )
+    team_id = Column(
+        ForeignKey("teams.id", onupdate="CASCADE", ondelete="CASCADE"),
+        nullable=False,
+    )
+
+    project = orm.relationship("Project", lazy=False)
+    team = orm.relationship("Team", lazy=False)
+
+
+class TeamFactory:
+    def __init__(self, request, organization=None):
+        self.request = request
+        self.organization = organization
+
+    def __getitem__(self, name):
+        if self.organization is None:
+            organization = OrganizationFactory(self.request)[name]
+            return TeamFactory(self.request, organization)
+        try:
+            return (
+                self.request.db.query(Team)
+                .filter(
+                    Team.organization == self.organization,
+                    Team.normalized_name == func.normalize_pep426_name(name),
+                )
+                .one()
+            )
+        except NoResultFound:
+            raise KeyError from None
+
+
+# TODO: Do we enable HasEvents functionality? Or are these just Org Events?
+# class Team(HasEvents, db.Model):
+class Team(db.Model):
+
+    __tablename__ = "teams"
+    __table_args__ = (
+        Index("teams_organization_id_idx", "organization_id"),
+        CheckConstraint(
+            r"name ~* '^([^\s/._-]|[^\s/._-].*[^\s/._-])$'::text",
+            name="teams_valid_name",
+        ),
+    )
+
+    __repr__ = make_repr("name", "organization")
+
+    name = Column(Text, nullable=False)
+    normalized_name = orm.column_property(func.normalize_team_name(name))
+    organization_id = Column(
+        ForeignKey("organizations.id", onupdate="CASCADE", ondelete="CASCADE"),
+        nullable=False,
+    )
+    created = Column(
+        DateTime(timezone=False),
+        nullable=False,
+        server_default=sql.func.now(),
+        index=True,
+    )
+
+    organization = orm.relationship("Organization", lazy=False, backref="teams")
+    # TODO: Or do we call this "members" instead?
```

Comment:
members is probably better in my opinion.

Reply:
Renamed to `members` in commit [11733f6](https://github.com/pypi/warehouse/pull/11665/commits/11733f688a8839adcb6dc39a4183a481c78b9cf5).